File: backend/src/040-expresiones_regulares/regexp.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# Referencia: https://docs.python.org/3.6/library/re.html

import re

texto = """Las expresiones regulares en Python se representan
con un carácter r delante de una cadena: """

expresion = r".a.."

# --------------------------------------------------

# Compilar expresiones regulares
patron = re.compile(expresion)
# https://docs.python.org/3.6/library/re.html#regular-expression-objects

# --------------------------------------------------

# Buscar todas las coincidencias (ejemplo compilando):
coincidencias = patron.findall(texto) # ['Las ', 'lare', 'cará', 'lant', 'na c']

# Ejemplo sin compilar:
coincidencias = re.findall(expresion, texto)

# --------------------------------------------------

# Buscar si hay coincidencias
coincidencia = re.search(expresion, texto) # <_sre.SRE_Match object; span=(1, 5), match='Las '>

# Retorna un objeto Match con la primera coincidencia si la hay o None si no
# https://docs.python.org/3.6/library/re.html#match-objects

# --------------------------------------------------

# Buscar coincidencia al principio de la cadena
coincidencia_al_principio = re.match(expresion, texto) # <_sre.SRE_Match object; span=(0, 4), match='Las '>

# Buscar coincidencia completa
coincidencia_completa = re.fullmatch(expresion, texto) # None

# --------------------------------------------------

# Substituir cadenas
substitucion = re.sub(expresion, "...", texto) # ...expresiones regu...s en Python se representan
                                               # con un ...cter r de...e de u...adena:

# --------------------------------------------------

File: backend/src/040-expresiones_regulares/regexp.js
"use strict"

var texto = `Las expresiones regulares en Javascript
se representan entre caracteres / `

var expresion = /.a../;

// Después de la expresión pueden aparecer modificadores:

var expresion2 = /.a../i,  // Coincidencias flexibles
    expresion3 = /.a../g,  // Buscar por todas las coincidencias,
                           //     sin detenerse en la primera
    expresion4 = /.a../m;  // Búqueda multilínea

// -----------------------------------------------

// Compilar expresiones regulares:
var patron = RegExp(expresion3);
// Podemos añadir como segundo parámetro los modificadores

// -----------------------------------------------

// Buscar por la primera coincidencia
var coincidencia = patron.exec(texto);

// Output:
//     [ 'Las ',
//     index: 0,
//     input: 'Las expresiones regulares en Javascript\nse representan entre caracteres / ' ]

// -----------------------------------------------

// Buscar si existe alguna coincidencia:
var coincidencia = patron.test(texto); // true

// -----------------------------------------------

// Buscar todas las coincidencias (con modificador g):
var coincidencias = texto.match(patron); // [ 'Las ', 'lare', 'Java', 'tan ', 'cara' ]

// -----------------------------------------------

// Índice de la primera coincidencia
var indice = texto.search(patron);  // 0

// -----------------------------------------------